### PR TITLE
Fix unstructured metadata accessors to respect omitempty semantics

### LINKED ISSUE #48211: Unstructured represents nil Object fields inconsistently
/kind bug

During the course of https://github.com/kubernetes/kubernetes/pull/48065, @ash2k [noted](https://github.com/kubernetes/kubernetes/pull/48065#issuecomment-311524304) that Unstructured is storing internal fields backing Object methods inconsistently with regards to nil. For example, SetOwnerReferences omits nil keys from the internal map, while SetInitializers, SetLabels, SetAnnotations, and SetDeletionTimestamp retain keys for nil values.

To remain consistent with marshalling behavior, we propose that Unstructured should omit keys for nil values internally for all fields which back Object methods.

/cc @ash2k @deads2k @caesarxuchao

### LINKED ISSUE #49075: SetAnnotations() could not distinguish  nil  and {}
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug
/sig api-machinery

**What happened**:
when I was working on pr https://github.com/kubernetes/kubernetes/pull/46560, I suspect it may be a bug with SetAnnotations(), since it cannot distinguish SetAnnotations(targetObj, nil) with SetAnnotations(targetObj, {}).

**What you expected to happen**:
SetAnnotations(targetObj, {})                  set   `annotations: {}`
SetAnnotations(targetObj, nil)                 remove  `annotations` field

**How to reproduce it (as minimally and precisely as possible)**:
see comments here : https://github.com/kubernetes/kubernetes/pull/46560#discussion_r127489967

### LINKED ISSUE #67541: k8s 1.11 custom object resource version changes on no-op status patch
> /kind bug

**What happened**:

In kubernetes 1.11.2 patching a status subresource of a custom object with a body that effectively does nothing increases the resource version of the object and triggers modify events.

In kubernetes 1.10.6 with the custom object subresources feature gate enabled this did not happen.

Maybe this is an intentional change due to the custom object subresources becoming beta, but I did not find this documented and it also does not not happen when patching the status of non custom objects (e.g. services) which is odd.

**What you expected to happen**:
Applying a patch that does not change the resource the object resource version should not change and no modify event be triggered.

**How to reproduce it (as minimally and precisely as possible)**:
Executing a merge patch (application/merge-patch+json) against a custom object changing the status multiple times will increase the resource version each time even when the object does not change.
E.g. via the python client:
api.patch_namespaced_custom_object_status(group=GROUP, version=version, plural=RESOURCE_NAME, namespace=namespace, name=object, body={"status": {"test": "data"}})

**Anything else we need to know?**:

**Environment**:
kubernetes version:
```
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.2", GitCommit:"bb9ffb1654d4a729bb4cec18ff088eacc153c239", GitTreeState:"clean", BuildDate:"2018-08-07T23:08:19Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```

### PR DESCRIPTION
**Pull request by nikhita**
Fixes #67541
Fixes #48211 
Fixes #49075
Follow up of #67562

`ObjectMeta` has fields with `omitempty` json tags. This means that when the fields have zero values, they should not be persisted in the object.

Before this PR, some of the metadata accessors for unstructured objects did not respect these semantics i.e they would persist a field even if it had a zero value.

This PR updates the accessors so that the field is removed from the unstructured object map if it contains a zero value.

/sig api-machinery
/kind bug
/area custom-resources
/cc sttts liggitt yue9944882 roycaihw 
/assign sttts liggitt 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by nikhita** [comment]
Does this need a "action required" release note?

**Comment by nikhita** [comment]
```
Some packages in hack/.golint_failures are passing golint. Please remove them.

  staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured
```
How is this even passing golint? There are so many linting errors in there. :woman_shrugging: 

/lint

**Comment by nikhita** [comment]
@cblecker @fejta re: golint and `/lint` above ^^

**Comment by sttts** [comment]
Looks good to me. @liggitt any further comments?

https://github.com/kubernetes/kubernetes/pull/67637 will add tests for the empty patch behaviour.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go [hunk 1]
```diff
@@ -220,31 +229,47 @@ func (u *Unstructured) GetName() string {
...
 	u.setNestedField(name, "metadata", "name")
 }
 
 func (u *Unstructured) GetGenerateName() string {
 	return getNestedString(u.Object, "metadata", "generateName")
 }
 
-func (u *Unstructured) SetGenerateName(name string) {
-	u.setNestedField(name, "metadata", "generateName")
+func (u *Unstructured) SetGenerateName(generateName string) {
```
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: well that's annoying

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go [hunk 2]
```diff
@@ -79,8 +79,8 @@ func (obj *Unstructured) ToList() (*UnstructuredList, error) {
 	return ret, nil
 }
 
-func (obj *Unstructured) EachListItem(fn func(runtime.Object) error) error {
-	field, ok := obj.Object["items"]
+func (u *Unstructured) EachListItem(fn func(runtime.Object) error) error {
```
[the remark this answers is not in the document]
Comment (nikhita) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go]: None of the methods in that file have comments and it doesn't really make sense to add to each one either. :/

**Comment by liggitt** [comment]
don't feed the linter... it'll just want another cookie :)

**Comment by nikhita** [comment]
> don't feed the linter... it'll just want another cookie :)

:)

I removed the last commit which tried to fix the golint errors. It's back to how it was before in the previous PR.

**Comment by nikhita** [comment]
Regarding the linting:

https://github.com/kubernetes/kubernetes/blob/816f2a48689247c8344d911995b0f97c721ad481/hack/verify-golint.sh#L75-L76

Looks like this completely ignores all errors if it sees a `foo_test` package.

**Comment by nikhita** [comment]
If golint sees a `foo_test` package in `foo`, it gives an error like:
`staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go is in package unstructured_test, not unstructured` and does not do any linting. We swallow this error and skip the linting.

https://github.com/kubernetes/kubernetes/blob/816f2a48689247c8344d911995b0f97c721ad481/hack/verify-golint.sh#L77

**Comment by liggitt** [comment]
this LGTM. @sttts, was there a way we were collecting dev-related release notes for client-go/apimachinery usage?

**Comment by nikhita** [comment]
> was there a way we were collecting dev-related release notes for client-go/apimachinery usage?

manually. :/
ref:
- https://gist.github.com/nikhita/3373e849f7a9dfcbc5e329121fdbfef1#file-how-to-generate-prs-md
- https://docs.google.com/spreadsheets/d/1jPexPgSohmD7zVmIeVvAFjjK_pqFu4MdjsmBXlINfvg/edit#gid=0
- https://github.com/kubernetes/test-infra/issues/5843

I was doing the manual release note process work for client-go anyway, so I'll take care of the release note for this one.

**Comment by nikhita** [comment]
Quick update here: there is a golint fix now - https://github.com/kubernetes/kubernetes/pull/67675 :tada:

**Comment by roycaihw** [comment]
This LGTM. One question: do we need to check zero values for `TypeMeta` (`SetKind` and `SetAPIVersion`) as well?

**Comment by liggitt** [comment]
> One question: do we need to check zero values for TypeMeta (SetKind and SetAPIVersion) as well?

 No valid serialized object has those set to empty values, so I don't think so

**Comment by sttts** [comment]
> was there a way we were collecting dev-related release notes for client-go/apimachinery usage

@liggitt we don't have prow support. But we usually use `dev-release-notes`

**Pull request merged**